ingester/ingester.go

```diff
@@ -0,0 +1,699 @@
+	}
+
+	i.queriedSamples.Add(float64(queriedSamples))
+
+	return result, nil
+}
+
+func samplesForRange(s *memorySeries, from, through model.Time) ([]model.SamplePair, error) {
+	// Find first chunk with start time after "from".
+	fromIdx := sort.Search(len(s.chunkDescs), func(i int) bool {
+		return s.chunkDescs[i].FirstTime().After(from)
+	})
+	// Find first chunk with start time after "through".
+	throughIdx := sort.Search(len(s.chunkDescs), func(i int) bool {
+		return s.chunkDescs[i].FirstTime().After(through)
+	})
+	if fromIdx == len(s.chunkDescs) {
+		// Even the last chunk starts before "from". Find out if the
+		// series ends before "from" and we don't need to do anything.
+		lt, err := s.chunkDescs[len(s.chunkDescs)-1].LastTime()
+		if err != nil {
+			return nil, err
+		}
+		if lt.Before(from) {
+			return nil, nil
+		}
+	}
+	if fromIdx > 0 {
+		fromIdx--
+	}
+	if throughIdx == len(s.chunkDescs) {
+		throughIdx--
+	}
+	var values []model.SamplePair
+	in := metric.Interval{
+		OldestInclusive: from,
+		NewestInclusive: through,
+	}
+	for idx := fromIdx; idx <= throughIdx; idx++ {
+		cd := s.chunkDescs[idx]
+		chValues, err := prom_chunk.RangeValues(cd.C.NewIterator(), in)
+		if err != nil {
+			return nil, err
+		}
+		values = append(values, chValues...)
+	}
+	return values, nil
+}
+
+// LabelValuesForLabelName returns all label values that are associated with a given label name.
+func (i *Ingester) LabelValuesForLabelName(ctx context.Context, name model.LabelName) (model.LabelValues, error) {
+	state, err := i.getStateFor(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return state.index.lookupLabelValues(name), nil
+}
+
+// Stop stops the Ingester.
+func (i *Ingester) Stop() {
+	i.stopLock.Lock()
+	i.stopped = true
+	i.stopLock.Unlock()
+
+	close(i.quit)
+	<-i.done
+}
+
+func (i *Ingester) loop() {
+	defer func() {
+		i.flushAllUsers(true)
+		close(i.done)
+		log.Infof("Ingester exited gracefully")
+	}()
+
+	tick := time.Tick(i.cfg.FlushCheckPeriod)
+	for {
+		select {
+		case <-tick:
+			i.flushAllUsers(false)
+		case <-i.quit:
+			return
+		}
+	}
+}
+
+func (i *Ingester) flushAllUsers(immediate bool) {
+	log.Infof("Flushing chunks... (exiting: %v)", immediate)
+	defer log.Infof("Done flushing chunks.")
+
+	if i.chunkStore == nil {
+		return
+	}
+
+	i.userStateLock.Lock()
+	userIDs := make([]string, 0, len(i.userState))
+	for userID := range i.userState {
+		userIDs = append(userIDs, userID)
+	}
+	i.userStateLock.Unlock()
+
+	var wg sync.WaitGroup
+	for _, userID := range userIDs {
+		wg.Add(1)
+		go func(userID string) {
+			i.flushUser(userID, immediate)
+			wg.Done()
+		}(userID)
+	}
+	wg.Wait()
+}
+
+func (i *Ingester) flushUser(userID string, immediate bool) {
+	log.Infof("Flushing user %s...", userID)
+	defer log.Infof("Done flushing user %s.", userID)
+
+	i.userStateLock.Lock()
+	userState, ok := i.userState[userID]
+	i.userStateLock.Unlock()
+
+	// This should happen, right?
+	if !ok {
+		return
```

Comment:
This seems like exactly the case for a `panic`: there's a non-recoverable programming error.